code/workspaces/infrastructure-api/src/kubernetes/namespaceApi.js

```diff
@@ -0,0 +1,74 @@
+import get from 'lodash/get';
+import { getCoreV1Api } from './kubeConfig';
+import logger from '../config/logger';
+
+export const MANAGED_BY = 'datalabs';
+
+async function readNamespaceMetadata(namespaceName) {
+  const k8sApi = getCoreV1Api();
+
+  try {
+    const response = await k8sApi.readNamespace(namespaceName);
+    const managedBy = get(response, 'body.metadata.labels.managedBy', undefined);
+    const managed = managedBy === MANAGED_BY;
+
+    return { namespace: namespaceName, exists: true, managed };
+  } catch (error) {
+    logger.debug(error);
+    return { namespace: namespaceName, exists: false, managed: false };
+  }
+}
+
+async function idempotentCreateNamespace(namespaceName) {
+  const k8sApi = getCoreV1Api();
+
+  // Namespace creation throws an error if the namespace exists so to ensure
+  // this is an idempotent operation We return early if the namespace exists
+  const namespaceMetadata = await readNamespaceMetadata(namespaceName);
+  if (namespaceMetadata.exists) {
+    return;
```

Comment:
Would it be useful to have a debug log message here?

Reply:
Yes, added